### PR TITLE
Add segmentation masking

### PR DESCRIPTION
**Pull request by gbrammer**
I've tried to implement some basic segmentation masking in the photometry and measurement functions (``flux_radius``, ``kron_radius``, but not yet ``winpos``) to address the point raised in  https://github.com/kbarbary/sep/issues/60.  It's not full ``MASK_TYPE = CORRECT`` behavior, but rather simply masks nonzero pixels in neighboring segments.

### CODE REVIEW

**Comment by kbarbary** [comment]
Nice! I'll take a closer look at this over the weekend. Could you add a test in `test.py`?

File: docs/apertures.rst [hunk 1]
```diff
@@ -161,6 +161,24 @@ of 0.5 and a normalizing flux of ``FLUX_AUTO``. The equivalent here is::
...
 
+Segmentation-masked image measurements
+--------------------------------------
+
+SourceExtractor provides a mechanism for measuring the "AUTO" and "FLUX_RADIUS" parameters for a given object including a mask for neighboring sources.  In addition to the mask, setting the SourceExtractor parameter ``MASK_TYPE=CORRECT`` further fills the masked pixels of a given source with "good" pixel values reflected opposite of the masked pixels.  The ``SEP`` photometry and measurement functions provide an option for simple masking without reflection or subtracting neighbor flux.  
+
+For example, using a segmentation array provided by ``sep.extract``, we can compute the masked ``flux_radius`` that could otherwise be artificially large due to flux from nearby sources::
+
+    # list of object id numbers that correspond to the segments
+    objs['id'] = np.arange(len(objs), dtype=np.int32)
```
Comment (kbarbary) [inline docs/apertures.rst]: In the segmentation array from `sep.extract`, the first object's pixels have value 1. So, I think this should be `np.arange(1, len(objs)+1, ...)`?

Comment (gbrammer) [reply docs/apertures.rst]: Yes, that's right.

**Comment by gbrammer** [comment]
I also just added a test, which, happily, actually works as expected!

I'm not sure why the travis `x.4` test is failing, but seems to be a more general numpy issue?

**Comment by kbarbary** [comment]
I suspect the travis failure is because of the absolutely ancient version of numpy (1.6) being tested. (Perhaps cython is generating code assuming numpy>=1.7). Don't worry about it here; the travis script should be updated to test only more recent numpy versions.

File: sep.pyx [hunk 2]
```diff
@@ -838,6 +857,16 @@ def sum_circle(np.ndarray data not None, x, y, r,
     maskthresh : float, optional
         Threshold for a pixel to be masked. Default is ``0.0``.
 
+    seg : `numpy.ndarray`, optional
+        Segmentation image with dimensions of `data` and dtype `np.float32`.
```
Comment (kbarbary) [inline sep.pyx]: dtype `np.int32`?

File: src/aperture.c [hunk 3]
```diff
@@ -430,17 +436,44 @@ int sep_sum_circann_multi(sep_image *im,
...
+
+    	           If `id` is positive, mask pixels with nonzero segment ids
+    	           not equal to `id`.
+
+    	      */ 
+    	      if (im->seg)
+      	        {
+      	          if (id > 0) 
+      	            {
+      	              if ((sconvert(segt) > 0.) & (sconvert(segt) != id))
```
Comment (kbarbary) [inline src/aperture.c]: Since this is in an inner loop, might be helpful for performance to only call `sconvert(segt)` once, and use short-circuiting `&&` instead of bitwise `&`.

File: sep.pyx [hunk 4]
```diff
@@ -838,6 +857,16 @@ def sum_circle(np.ndarray data not None, x, y, r,
...
+    
+    seg_id : array_like, optional
+        Array of segmentation ids that correspond to the dimensions of `x` 
+        and `y`.  If the values in seg_id are negative, then the mask is 
+        generated requiring pixels within the segment.  Otherwise, mask 
+        pixels in `seg` with nonzero values different from `seg_id` for a 
+        given object in the list.  Must be provided along with `seg`.
```
Comment (kbarbary) [inline sep.pyx]: I found this a bit hard to parse the first time I read it. How about
```
Array of segmentation ids  used to mask additional pixels in the image.
Dimensions correspond to the dimensions of ``x`` and ``y``. The behavior differs
depending on whether ``seg_id`` is negative or positive. If ``seg_id`` is positive,
all pixels belonging to other objects are masked. (Pixel ``j, i`` is masked if
``seg[j, i] != seg_id and seg[j, i] != 0``). If ``seg_id`` is negative, all pixels
other than those belonging to the object of interest are masked. (Pixel ``j, i`` is
masked if ``seg[j, i] != -seg_id``).
```
Note the double backticks so that this renders correctly in restructured text docs (rather than the markdown single-backticks).

File: sep.pyx [hunk 5]
```diff
@@ -873,8 +902,13 @@ def sum_circle(np.ndarray data not None, x, y, r,
     cdef int status
     cdef np.broadcast it
     cdef sep_image im
-
-    _parse_arrays(data, err, var, mask, &im)
+            
+    # Test for seg without seg_id.  Nothing happens if seg_id supplied but 
+    # without seg.
+    if (seg is not None) & (seg_id is None):
```
Comment (kbarbary) [inline sep.pyx]: This is very minor, but probably better to use `and` instead of `&` here.

File: sep.pyx [hunk 6]
```diff
@@ -15,7 +15,7 @@ from cpython.version cimport PY_MAJOR_VERSION
 
 np.import_array()  # To access the numpy C-API.
 
-__version__ = "1.0.3"
+__version__ = "1.0.4"
```
Comment (kbarbary) [inline sep.pyx]: You can make this 1.1.0 since this is a new feature, not just a bug fix.

**Comment by kbarbary** [comment]
Really excellent work on this PR! Thanks!

File: sep.pyx (hunk 2 above)
Comment (gbrammer) [reply sep.pyx]: Agreed.  I just renamed "seg" to "segmap" throughout and updated the docstrings to make the direct connection to the output from `sep.extract` more clear.

File: sep.pyx (hunk 4 above)
Comment (gbrammer) [reply sep.pyx]: Agreed.  Docstrings updated in 71328de0b2bf590b2fa0f5a62eb15095b89863a6.

File: sep.pyx (hunk 5 above)
Comment (gbrammer) [reply sep.pyx]: Fair enough (f25b42d7f10323b1d598e6aa65037a7c8dbd14bb).  I never use `and` but perhaps I should....

File: sep.pyx (hunk 6 above)
Comment (gbrammer) [reply sep.pyx]: Done! (c5e3dedbd0797a2b392f899f71eed69790e711df).

**Comment by gbrammer** [comment]
Happy to help.  This is an excellent package you've made and makes life
with(out) SExtractor so much easier!

File: sep.pyx (hunk 5 above)
Comment (kbarbary) [reply sep.pyx]: Yeah, definitely not a big deal. Generally the benefit of `and` is that it is short-circuiting. Here that is irrelevant (`seg_id is None` takes very little time anyway), so the only benefit is clarity of intent: so the reader isn't asking "wait, why is a bitwise operation necessary here?"

**Comment by gbrammer** [comment]
Just realized I messed up the version to be "v1.10.0" instead of "v1.1.0"!  Sorry about that.
